diff --git a/Makefile b/Makefile
index 74e454f..24e6309 100644
--- a/Makefile
+++ b/Makefile
@@ -194,7 +194,10 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
-
+	$U/_pingpong\
+	$U/_primes\
+	$U/_find\
+	$U/_xargs\
 
 
 
diff --git a/user/find.c b/user/find.c
new file mode 100644
index 0000000..22faa24
--- /dev/null
+++ b/user/find.c
@@ -0,0 +1,100 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/fs.h"
+#include "kernel/fcntl.h"
+
+char *fmtname(char *path)
+{
+    char *p = path + strlen(path);
+    while (p >= path && *p != '/')
+        p--;
+    return p + 1;
+}
+
+void find(char *path, char *file)
+{
+    char buffer[512], *p;
+    int fd;
+    struct dirent d;
+    struct stat s;
+
+    fd = open(path, O_RDONLY);
+    if (fd < 0)
+    {
+        fprintf(2, "Error: cannot open %s\n", path);
+        return;
+    }
+
+    if (fstat(fd, &s) < 0)
+    {
+        fprintf(2, "Error: cannot stat %s\n", path);
+        close(fd);
+        return;
+    }
+
+    switch (s.type)
+    {
+    case T_FILE:
+        if (strcmp(fmtname(path), file) == 0)
+        {
+            printf("%s\n", path);
+        }
+        break;
+
+    case T_DIR:
+        if (strlen(path) + 1 + DIRSIZ + 1 > sizeof(buffer))
+        {
+            fprintf(2, "Error: path too long\n");
+            break;
+        }
+
+        strcpy(buffer, path);
+        p = buffer + strlen(buffer);
+        *p++ = '/';
+
+        while (read(fd, &d, sizeof(d)) == sizeof(d))
+        {
+            if (d.inum == 0)
+                continue;
+            if (strcmp(d.name, ".") != 0 && strcmp(d.name, "..") != 0)
+            {
+                memmove(p, d.name, DIRSIZ);
+                p[DIRSIZ] = 0;
+
+                if (stat(buffer, &s) < 0)
+                {
+                    fprintf(2, "Error: cannot stat %s\n", buffer);
+                    continue;
+                }
+                find(buffer, file);
+            }
+        }
+        break;
+    }
+
+    close(fd);
+}
+
+int main(int argc, char *argv[])
+{
+    if (argc < 2)
+    {
+        fprintf(2, "Usage: find path file\n");
+        exit(1);
+    }
+
+    if (argc == 2)
+    {
+        find(".", argv[1]);
+    }
+    else
+    {
+        for (int i = 1; i < argc - 1; i++)
+        {
+            find(argv[i], argv[argc - 1]);
+        }
+    }
+
+    exit(0);
+}
diff --git a/user/pingpong.c b/user/pingpong.c
new file mode 100644
index 0000000..c986c4b
--- /dev/null
+++ b/user/pingpong.c
@@ -0,0 +1,59 @@
+#include "kernel/types.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[])
+{
+    int parent_to_child_pipe[2], child_to_parent_pipe[2];  
+    char byte = 'a';   
+    char buffer[1];   
+    
+    if (pipe(parent_to_child_pipe) < 0 || pipe(child_to_parent_pipe) < 0) {
+        fprintf(2, "Error: Can't create pipe!\n");
+        exit(1);
+    }
+
+    int pid = fork();  
+
+    if (pid < 0) {  
+        fprintf(2, "Error: Can't fork!\n");
+        exit(1);
+    }
+
+    if (pid == 0) {  
+        
+        close(parent_to_child_pipe[1]);  
+        close(child_to_parent_pipe[0]);  
+
+        
+        if (read(parent_to_child_pipe[0], buffer, 1) > 0 && buffer[0] == 'a') {
+            printf("%d: received ping\n", getpid()); 
+        }
+
+        
+        write(child_to_parent_pipe[1], buffer, 1);
+
+        
+        close(parent_to_child_pipe[0]); 
+        close(child_to_parent_pipe[1]);  
+        exit(0);  
+    } else {  
+        
+        close(parent_to_child_pipe[0]);  
+        close(child_to_parent_pipe[1]); 
+
+        
+        write(parent_to_child_pipe[1], &byte, 1);
+
+        
+        if (read(child_to_parent_pipe[0], buffer, 1) > 0 && buffer[0] == 'a') {
+            printf("%d: received pong\n", getpid()); 
+        }
+
+        
+        close(parent_to_child_pipe[1]); 
+        close(child_to_parent_pipe[0]); 
+
+        wait(0);
+        exit(0); 
+    }
+}
diff --git a/user/primes.c b/user/primes.c
new file mode 100644
index 0000000..2f571c8
--- /dev/null
+++ b/user/primes.c
@@ -0,0 +1,81 @@
+#include "kernel/types.h"
+#include "user/user.h"
+
+void primes(int left_pipe)
+{
+    int prime;
+
+    while (1) {
+        if (read(left_pipe, &prime, sizeof(prime)) == 0) {
+            close(left_pipe);
+            exit(0); 
+        }
+
+        printf("prime %d\n", prime);
+
+        int num;
+        int right_pipe[2];
+
+        if (pipe(right_pipe) < 0) {
+            fprintf(2, "Error: cannot create pipe\n");
+            exit(1);
+        }
+
+        int pid = fork();
+        if (pid < 0) {
+            fprintf(2, "Error: fork failed\n");
+            exit(1);
+        } else if (pid == 0) {
+            close(left_pipe);          
+            close(right_pipe[1]);      
+            left_pipe = right_pipe[0]; 
+        } else {
+            close(right_pipe[0]);      
+
+            while (read(left_pipe, &num, sizeof(num)) > 0) {
+                if (num % prime != 0) {
+                    write(right_pipe[1], &num, sizeof(num));  
+                }
+            }
+
+            close(left_pipe);
+            close(right_pipe[1]);
+
+            wait(0);
+            exit(0);
+        }
+    }
+}
+
+int main(int argc, char *argv[])
+{
+    int initial_pipe[2];
+
+    if (pipe(initial_pipe) < 0) {
+        fprintf(2, "Error: cannot create initial pipe\n");
+        exit(1);
+    }
+
+    int pid = fork();
+    if (pid < 0) {
+        fprintf(2, "Error: fork failed\n");
+        exit(1);
+    } else if (pid == 0) {
+        close(initial_pipe[1]);    
+        primes(initial_pipe[0]);  
+    } else {
+        close(initial_pipe[0]);   
+
+        
+        for (int i = 2; i <= 280; i++) {
+            write(initial_pipe[1], &i, sizeof(i));
+        }
+
+        
+        close(initial_pipe[1]);
+
+        
+        wait(0);
+        exit(0);
+    }
+}
diff --git a/user/xargs.c b/user/xargs.c
new file mode 100644
index 0000000..87fc2a6
--- /dev/null
+++ b/user/xargs.c
@@ -0,0 +1,96 @@
+#include "kernel/types.h"
+#include "kernel/param.h"
+#include "user/user.h"
+
+int main(int argc, char* argv[]) {
+    // Các tham số cơ bản khi không có -n
+    int argPerCommand = MAXARG;
+    int cmdStart = 1;
+    
+    // Xử lý trường hợp có tham số -n
+    if (strcmp(argv[1], "-n") == 0) {
+        argPerCommand = atoi(argv[2]);
+        cmdStart = 3;
+    }
+
+    // Lưu các tham số đầu vào của xargs
+    char* args[MAXARG];
+    char buf[1024];
+    int n = 0;
+    while (n < argc - cmdStart) {
+        args[n] = argv[cmdStart + n];
+        n++;
+    }
+
+    // Thực hiện 1 lệnh
+    while (1) {
+        int i = 0;
+        int status = 0;
+
+        // Đọc các tham số từ đầu vào chuẩn
+        while (1) {
+            status = read(0, &buf[i], 1);
+            if (status < 1) {
+                break;
+            }
+            if (buf[i] == '\n') break;
+            i++;
+        }
+        
+        // Không còn dữ liệu đầu vào (EOF)
+        if (i == 0 && status < 1) break;
+
+        // Xử lý với -n tham số cho mỗi lệnh
+        int argsIdx = n;
+        int startArgsIdx = 0;
+        int argsCount = 0;
+        int j = 0;
+        while (j <= i) {
+            // Tách các tham số từ đâu vào chuẩn
+            if (buf[j] == ' ' || buf[j] == '\n') {
+                buf[j] = 0;
+                args[argsIdx++] = &buf[startArgsIdx];
+                startArgsIdx = j + 1;
+                argsCount++;
+            }
+            // Đủ tham số theo -n thì thực hiện
+            if (argsCount == argPerCommand) {
+                args[argsIdx] = 0;
+                // Tạo ra tiến trình còn và thực hiện lệnh
+                if (fork() == 0) {
+                    exec(args[0], args);
+
+                    // Nếu lệnh exec fail
+                    fprintf(2, "exec %s failed\n", args[0]);
+                    exit(1);
+                }
+                else {
+                    wait(0);
+                }
+                argsIdx = n;
+                argsCount = 0;
+            }
+            j++;
+        }
+
+        // Trường hợp còn thừa nhưng không đủ tham số của -n
+        if (argsCount != 0) {
+            args[argsIdx] = 0;
+            // Tạo ra tiến trình còn và thực hiện lệnh
+            if (fork() == 0) {
+                exec(args[0], args);
+
+                // Nếu lệnh exec fail
+                fprintf(2, "exec %s failed\n", args[0]);
+                exit(1);
+            }
+            else {
+                wait(0);
+            }
+            argsIdx = n;
+            argsCount = 0;
+        }
+    }
+    // Kết thúc tiến trình hoàn thành
+    exit(0);
+}
\ No newline at end of file
